packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -379,37 +336,29 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
     function delegateKghBatch(
         address validator,
         uint256[] calldata tokenIds
-    ) external isRegistered(validator) returns (uint128, uint128) {
+    ) external isRegistered(validator) returns (uint128) {
         if (tokenIds.length == 0) revert NotAllowedZeroInput();
 
+        // TODO: claim boosted reward and delegate to KRO pool
+
         uint128 kroShares;
-        uint128 kghShares = _convertToKghShares(validator);
         uint128 kroInKghs;
 
-        for (uint256 i = 0; i < tokenIds.length; ) {
-            KGH.safeTransferFrom(msg.sender, address(this), tokenIds[i]);
-
-            uint128 kroInKgh = KGH_MANAGER.totalKroInKgh(tokenIds[i]);
-            uint128 kroSharesForTokenId = _convertToKroShares(validator, kroInKgh);
-
-            _vaults[validator].kghDelegators[msg.sender].shares[tokenIds[i]] = KghDelegatorShares({
-                kro: kroSharesForTokenId,
-                kgh: kghShares
-            });
-
-            unchecked {
-                kroInKghs += kroInKgh;
-                kroShares += kroSharesForTokenId;
+        Asset storage asset = _vaults[validator].asset;
+        KghDelegator storage delegator = _vaults[validator].kghDelegators[msg.sender];
 
-                ++i;
-            }
+        for (uint256 i = 0; i < tokenIds.length; ) {
+            (kroInKgh, kroShare) = _delegateKgh(validator, asset, delegator, tokenIds[i]);
```

Comment:
Do you mean `_delegateKghBatch`? Please check the function if it's right.
```suggestion
            (uint128 kroInKgh, uint128 kroShare) = _delegateKgh(validator, asset, delegator, tokenIds[i]);
```

Reply:
If you meant to use `_delegateKgh` here also, I think it would be inefficient in the respect of gas. We don't need to update storage every time when using `_delegateKghBatch`.

Reply:
Yes I meant, but you're right. It is gas inefficient when many KGHs are delegated at once. I'll rollback to use `_delegetKghBatch`